File: 14/main.py
# divide and conquer

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if strs == None or len(strs) <= 0:
            return ""
        
        return self.longestCommonPrefixHelper(strs, 0, len(strs) - 1)
    
    def longestCommonPrefixHelper(self, strs, left, right):
        if left == right:
            return strs[left]
        else:
            mid = (left + right) // 2
            str1 = self.longestCommonPrefixHelper(strs, left, mid)
            str2 = self.longestCommonPrefixHelper(strs, mid + 1, right)
            return self.commonPrefix(str1, str2)
        
    def commonPrefix(self, str1, str2):
        minimum = min(len(str1), len(str2))
        result = ""
        
        for i in range(minimum):
            if str1[i] != str2[i]:
                result = str1[:i]
                return result
            
        result = str1[:minimum]
        return result
        

File: 14/main.java
// divide and conquer

class Solution {
    public String longestCommonPrefix(String[] strs) {
        if(strs == null || strs.length <= 0) {
            return "";
        }
        
        return longestCommonPrefixHelper(strs, 0, strs.length - 1);
    }
    
    private String longestCommonPrefixHelper(String[] strs, int left, int right) {
        if(left == right) {
            return strs[left];
        } else {
            int mid = (left + right) / 2;
            String str1 = longestCommonPrefixHelper(strs, left, mid);
            String str2 = longestCommonPrefixHelper(strs, mid + 1, right);
            return commonPrefix(str1, str2);
        }
    }
    
    private String commonPrefix(String str1, String str2) {
        int minimum = Math.min(str1.length(), str2.length());
        String result;
        
        for(int i = 0; i < minimum; i++) {
            if(str1.charAt(i) != str2.charAt(i)) {
                result = str1.substring(0, i);
                return result;
            }
        }
        
        result = str1.substring(0, minimum);
        return result;
    }
}
